jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -0,0 +1,407 @@
+                total_loss, grads = grad_fn(
+                    params, opt_state, traj_batch, advantages, targets
+                )
+
+                grads, total_loss = jax.lax.pmean(
+                    (grads, total_loss), axis_name="batch"
+                )
+                grads, total_loss = jax.lax.pmean(
+                    (grads, total_loss), axis_name="device"
+                )
+
+                updates, new_opt_state = update_fn(grads, opt_state)
+                new_params = optax.apply_updates(params, updates)
+
+                return (new_params, new_opt_state), total_loss
+
+            params, opt_state, traj_batch, advantages, targets, rng = update_state
+            rng, _rng = jax.random.split(rng)
+
+            # SHUFFLE MINIBATCHES
+            batch_size = config["ROLLOUT_LENGTH"] * config["NUM_ENVS"]
+            permutation = jax.random.permutation(_rng, batch_size)
+            batch = (traj_batch, advantages, targets)
+            batch = jax.tree_util.tree_map(
+                lambda x: x.reshape((batch_size,) + x.shape[2:]), batch
+            )
+            shuffled_batch = jax.tree_util.tree_map(
+                lambda x: jnp.take(x, permutation, axis=0), batch
+            )
+            minibatches = jax.tree_util.tree_map(
+                lambda x: jnp.reshape(
+                    x, [config["NUM_MINIBATCHES"], -1] + list(x.shape[1:])
+                ),
+                shuffled_batch,
+            )
+
+            # UPDATE MINIBATCHES
+            (params, opt_state), total_loss = jax.lax.scan(
+                _update_minibatch, (params, opt_state), minibatches
+            )
+
+            update_state = (params, opt_state, traj_batch, advantages, targets, rng)
+            return update_state, total_loss
+
+        update_state = (params, opt_state, traj_batch, advantages, targets, rng)
+
+        # UPDATE EPOCHS
+        update_state, loss_info = jax.lax.scan(
+            _update_epoch, update_state, None, config["PPO_EPOCHS"]
+        )
+
+        params, opt_state, traj_batch, advantages, targets, rng = update_state
+        runner_state = (params, opt_state, rng, env_state, last_timestep)
+        metric = traj_batch.info
+        return runner_state, metric
+
+    def learner_fn(
+        params: FrozenDict,
+        opt_state: OptState,
+        rng: chex.PRNGKey,
+        env_state: State,
+        timesteps: TimeStep,
+    ) -> dict:
+        """Learner function."""
+        runner_state = (params, opt_state, rng, env_state, timesteps)
+
+        batched_update_step = jax.vmap(
+            _update_step, in_axes=(0, None), axis_name="batch"
+        )
+
+        runner_state, metric = jax.lax.scan(
+            batched_update_step, runner_state, None, config["NUM_UPDATES"]
+        )
+        return {"runner_state": runner_state, "metrics": metric}
+
+    return learner_fn
+
+
+def run_experiment(env: Environment, config: dict) -> dict:
```

Comment:
In `types.py` we should have a config type something like `Dict[str, Union[chex.Array, str]`. Also could we update the return type from `dict` to `Dict[...]`

Reply:
for now I guess we should keep the config as dict and later on we will change it as it will be updated in the next PR where we will add the logger and evaluator